ext/bcmath/libbcmath/src/div.c

```diff
@@ -31,219 +31,434 @@
+	 * MAX_E = (10^(x + b) * (10^(k * b) - 1)) / (10^(x + k * b) * (10^x + 1) + 10^x)
+	 * MAX_E < (10^(x + b) * 10^(k * b)) / (10^(x + k * b) * 10^x + 10^x)
+	 * MAX_E < (10^(x + k + 2b)) / (10^x * (10^(x + k * b) + 1))
+	 *
+	 * Approximate again and remove +1 in the denominator
+	 * MAX_E < 10^(x + k + 2b) / (10^x * (10^(x + k * b) + 1))
+	 * MAX_E < 10^(x + k + 2b) / (10^x * 10^(x + k * b))
+	 * MAX_E < 10^b / 10^x
+	 * MAX_E < 10^(b - x)
+	 *
+	 * Therefore, found that if the number of digits in radix B and n2_high are equal, the error will be less
+	 * than 1 regardless of the number of digits in the value of k.
+	 *
+	 * Here, B is BC_VECTOR_BOUNDARY_NUM, so adjust the number of digits in divider to be BC_VECTOR_SIZE + 1.
+	 */
+	size_t n2_top_digits = n2_len % BC_VECTOR_SIZE;
+	if (n2_top_digits == 0) {
+		n2_top_digits = BC_VECTOR_SIZE;
+	}
+	size_t high_part_shift = POW_10_LUT[BC_VECTOR_SIZE - n2_top_digits + 1];
+	size_t low_part_shift = POW_10_LUT[n2_top_digits - 1];
+	BC_VECTOR divider = n2_vector[n2_top_index] * high_part_shift + n2_vector[n2_top_index - 1] / low_part_shift;
+	for (i = 0; i < quot_arr_size; i++) {
+		BC_VECTOR divided = n1_vector[n1_top_index - i] * high_part_shift + n1_vector[n1_top_index - i - 1] / low_part_shift;
+
+		/* If it is clear that n2 is greater in this loop, then the quotient is 0. */
+		if (div_carry == 0 && divided < divider) {
+			quot_vector[quot_top_index - i] = 0;
+			div_carry = n1_vector[n1_top_index - i];
+			n1_vector[n1_top_index - i] = 0;
+			continue;
+		}
 
-			if (carry != 0) {
-				*rptr = carry;
+		/*
+		 * Determine the temporary quotient.
+		 * "divided" is n1_high in the previous example. The maximum value of n1_high is n2_high * B,
+		 * so here it is divider * BC_VECTOR_BOUNDARY_NUM.
+		 * The number of digits in divider is BC_VECTOR_SIZE + 1, so even if divider is at its maximum value,
+		 * it will never overflow here.
+		 */
+		divided += div_carry * BC_VECTOR_BOUNDARY_NUM * high_part_shift;
+		BC_VECTOR quot_guess = divided / divider;
+
+		/* For sub, add the remainder to the high-order digit */
+		n1_vector[n1_top_index - i] += div_carry * BC_VECTOR_BOUNDARY_NUM;
+
+		/*
+		 * The temporary quotient can only have errors in the "too large" direction.
+		 * So if the temporary quotient is 0 here, the quotient is 0.
+		 */
+		if (quot_guess == 0) {
+			quot_vector[quot_top_index - i] = 0;
+			div_carry = n1_vector[n1_top_index - i];
+			n1_vector[n1_top_index - i] = 0;
+			continue;
+		}
+
+		/* sub */
+		BC_VECTOR sub;
+		BC_VECTOR borrow = 0;
+		BC_VECTOR *n1_calc_bottom = n1_vector + n1_arr_size - n2_arr_size - i;
+		for (j = 0; j < n2_arr_size - 1; j++) {
+			sub = n2_vector[j] * quot_guess + borrow;
+			BC_VECTOR sub_low = sub % BC_VECTOR_BOUNDARY_NUM;
+			borrow = sub / BC_VECTOR_BOUNDARY_NUM;
+
+			if (n1_calc_bottom[j] >= sub_low) {
+				n1_calc_bottom[j] -= sub_low;
+			} else {
+				n1_calc_bottom[j] += BC_VECTOR_BOUNDARY_NUM - sub_low;
+				borrow++;
```

Comment:
Why is the borrow never reset? I think I'm missing something here.

Reply:
The value is always overwritten with `borrow = sub / BC_VECTOR_BOUNDARY_NUM` in L209, so it's okay.